feat(portal): Application <> consumer mapping (#5500)
Kong/developer.konghq.com: app/dev-portal/self-service.md

* API Packages have a per-request PATCH limit of 100.

## Apply plugins to applications {% new_in 3.15 %}

You can apply {{site.base_gateway}} plugins to your {{site.dev_portal}} applications.
This lets you enforce business logic, such as rate limiting or IP restriction, on the credentials that an application uses to access your APIs.

The following table shows common use cases for applying plugins to applications:

<!--vale off-->
{% table %}
columns:
  - title: Use case
    key: use_case
  - title: Plugin
    key: plugin
rows:
  - use_case: "Enforce request quotas on the credentials an application uses."
    plugin: "[Rate Limiting](/plugins/rate-limiting/) or [Rate Limiting Advanced](/plugins/rate-limiting-advanced/)"
  - use_case: "Ensure only requests from a partner's known IP ranges can use their application credentials."
    plugin: "[IP Restriction](/plugins/ip-restriction/)"
  - use_case: "Automatically inject a header identifying the partner into every request their application makes, so your upstream can route or log by customer without trusting client-supplied headers."
    plugin: "[Request Transformer](/plugins/request-transformer/)"
{% endtable %}
<!--vale on-->

> When a developer creates an application, {{site.dev_portal}} automatically adds a Kong Identity principal to the application.

Reviewer: ```suggestion
Plugins can be applied to an application by linking the application to an existing consumer in API Gateway. This linking is done automatically via Kong Identity principals. When a developer creates an application, {{site.dev_portal}} automatically creates a Kong Identity principal for this application. This principal entity then helps link the application to an existing consumer entity in API Gateway.

Kong Identity does not store any credentials for applications, it is just used for mapping an application to a consumer.
```

Reviewer: This may be way too much internal details for our users. Happy to collaborate to figure out what is the right level of detail that we provide while introducing the role of Kong Identity in this.

Author: @smritikjaggi Do we want to specify that there are two methods to apply a plugin to an application? So something more like:

"Plugins can be applied to an application by either linking the application to an existing Consumer in Kong Gateway or applying the plugin directly to the associated principal."

Reviewer 2: With the wording for applying plugins to principals, there is not a mechanism to apply a plugin "directly" to a principal (i.e., through a direct principal to plugin relationship) in the same way that plugins can be applied to entities like consumers, routes, and services. Conditional execution logic can be used to target the specific principal, but this is an indirect and loosely coupled way of associating plugins with principals.

You can apply plugins to an application in two different ways:

> * **Conditional plugin execution (recommended):** Use [conditional plugin execution](/gateway/configure-conditional-plugin-execution/) with an expression that references the application's `principal.id`. This applies the plugin based on the application's principal, without mapping the application to a Consumer.

Reviewer: ```suggestion
* **Conditional plugin execution (recommended):** Use [conditional plugin execution](/gateway/configure-conditional-plugin-execution/) with an expression that references the application's `principal.id`. This applies the plugin based on the application's principal, without mapping the application to a Consumer. This configuration can be managed within Kong Identity and API Gateway
```

Reviewer: I am hesitant to call this approach recommended as it is valuable for only net new customers. Our existing customers all fall in the second bucket for consumer scoped plugins because they started their journey in API Gateway and then adopted Dev Portal. 

It is also likely that net new customers also start their journey in API Gateway and adopt Dev Portal after, making them fall in the second bucket of consumer scoped plugin. Gateway forms our entry point overall in Konnect.

Reviewer: After discussing with Mike, another way to write this is - "if you are starting net new and have not created consumers, you can directly apply plugins to the principal".

Author: I'll turn this into a table so it's easier to convey this info

* **Consumer-scoped plugins:** [Map the application to an existing Gateway Consumer](#map-an-application-to-a-consumer), then configure Consumer-scoped plugins on that Consumer. This is a common starting point if you already have Consumers configured.

> ### Apply a plugin to an application using a principal

Reviewer: ```suggestion
### Apply a plugin to an application using a Kong Identity Principal
```

status_code: 204
region: us
method: PUT
body:
  id: $CONSUMER_ID
{% endkonnect_api_request %}
<!--vale on-->

{% endnavtab %}
{% navtab "UI" %}
1. In the {{site.konnect_short_name}} sidebar, click **Dev Portal > Portals**.
1. Click your portal.
1. Click the **Access and approvals** tab.
1. Click the **App Registrations** tab.
1. Click the application you want to link a Consumer to.
1. In the **App Registrations** section, click the action menu icon for the registration.
1. Click **Link Consumer**.
1. From the **Consumer** dropdown menu, select the Consumer you want to link.
   The API must be [linked to a Gateway Service or control plane](/catalog/apis/#allow-developers-to-consume-your-api) to link a Consumer.
1. Click **Link consumer**.
{% endnavtab %}
{% endnavtabs %}

Any plugins that were applied to the Consumer are now applied to the {{site.dev_portal}} application.

Keep the following in mind when you map applications to Consumers:
* Both the [KAA and ACE plugins](/catalog/apis/#allow-developers-to-consume-your-api) look up principals to resolve the Consumer mapped to an application.
* An application maps to a single Consumer (a 1:1 mapping through one principal).

> * The mapping is a loose, string-based mapping. The {{site.dev_portal}} does not validate that the Consumer exists on the Gateway.

Reviewer: This is incorrect. Mapping is done via consumer ID, so the consumer must exist before mapping can be done successfully. We should confirm if the API will give a 404 or bad request for an incorrect consumer ID. I know that the UI does not allow to map to a consumer that does not exist.

Reviewer 2: The response for an incorrect consumer ID will be a 404

* Application registrations for APIs that are linked to the same Gateway Service will share the same effective Consumer mapping.
  Updating the mapping for one registration updates it for all registrations that resolve to the same Gateway context.

> * Applying plugins to applications is only available on v3 {{site.dev_portal}}s.

Reviewer: This is true for the conditional execution approach also, lets add it in that section too.

Author: I'll break this section into it's own h3 so it will apply to both.

<!--vale off-->
{% konnect_api_request %}
url: /v3/portals/$PORTAL_ID/applications?filter%5Bname%5D%5Beq%5D=$APPLICATION_NAME
status_code: 200
region: us
method: GET
capture:
  - variable: PRINCIPAL_ID
    jq: '.data[0].id'
{% endkonnect_api_request %}
<!--vale on-->
   The principal ID is the same as the application's ID. 
1. Configure the Rate Limiting Advanced plugin and use a conditional plugin execution expression to apply it to the application's principal. Replace `$CONTROL_PLANE_ID` with the ID of the control plane that your API is linked to:
<!--vale off-->
{% konnect_api_request %}
url: /v2/control-planes/$CONTROL_PLANE_ID/core-entities/plugins/
status_code: 201
region: us
method: POST
body:
  name: rate-limiting-advanced
  config:
    limit:
      - 200
    window_size:
      - 1800
    window_type: fixed
    namespace: my-namespace
  condition: 'principal.id == "$PRINCIPAL_ID"'
{% endkonnect_api_request %}
<!--vale on-->

Any request that authenticates as this application is now rate limited to 200 requests every 30 minutes.

### Map an application to a Consumer

> Mapping an application to a Consumer requires the [Consumer Viewer](/konnect-platform/teams-and-roles/#control-planes) role, granted for each API instance registered by the application.

Author: I've asked about this in the project channel
